### PR TITLE
Dedicate channel handlers & Subscriber context & Subscriber pool & Sa…

### PR DESCRIPTION
**Pull request by allenlee820202**
…mpleSubscribe client

Dedicate channel handlers:
1. MessageParseHandler parses byte stream into Messages
2. ConnectHandler concatenates after the MessageParseHandler, it checks the user
information and add the third handler at pipeline().last. The third handler can
either be PublishHandler or SubscribeHandler
3. PublishHandler puts the published message into InMemoryPool
4. SubscribHandler registers the subscriber into SubscriberPool

Subscriber context
The subscriber context records
1. The topic
2. The groupId
3. The channel handler context

Subscriber pool
1. Store A map with topic key to a map of subscribers.
2. The map of subscribers maps the ChannelHandlerContext instance to a
subscriber context.
3. The SubscribeHandler registers the subscriber with its channel
handler context as key and its subscriber context as value

SampleSubscribe client
1. The subscribe client for demo

@tz70s Please help take a look, thanks!

### CODE REVIEW

File: broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessageParseHandler.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        ByteBuf incoming = (ByteBuf)msg;
+        StringBuilder packet = new StringBuilder();
+        ArrayDeque<Message> newMessages = new ArrayDeque();
+
+        try {
+            while(incoming.isReadable()) {
+                packet.append((char)incoming.readByte());
+            }
+
```
Comment (tz70s) [inline broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessageParseHandler.java]: You may further investigate the message decoder/encoder mechanism in the future.

File: broke-broker/src/main/java/org/dsngroup/broke/broker/storage/InMemoryPool.java [hunk 2]
```diff
@@ -27,16 +28,24 @@
...
      */
     public static void putContentOnTopic(String topic, String content) {
+
+        // Initialize the array list for the first-time topic
+        if (inMemoryPool.get(topic) == null)
+            inMemoryPool.put(topic, new ArrayList());
+
         // Ignore the return value
-        inMemoryPool.put(topic, content);
+        inMemoryPool.get(topic).add(content);
```
Comment (tz70s) [inline broke-broker/src/main/java/org/dsngroup/broke/broker/storage/InMemoryPool.java]: In my guess, the `ArraList` here is used for concat the content under associated topic?
Try for offset or partition?

File: broke-broker/src/main/java/org/dsngroup/broke/broker/storage/SubscriberPool.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        SubscriberContext subscriber = new SubscriberContext(topic, groupId, channelCxt);
+
+        subscriberPool.get(topic).put(channelCxt, subscriber);
+
+        // TODO: remove this test
+        // System.out.println( "Test the subscriber pool: \n"+subscriberPool.get(topic).get(channelCxt).getClass().toString() );
+        return subscriber;
+    }
+
```
Comment (tz70s) [inline broke-broker/src/main/java/org/dsngroup/broke/broker/storage/SubscriberPool.java]: I think we can also consider how to generate a more well-defined data structure / classes between `SubscriberPool` and `InMemoryPool`, in the future.

File: broke-broker/src/main/java/org/dsngroup/broke/broker/storage/InMemoryPool.java (hunk 2 above)
Comment (allenlee820202) [reply broke-broker/src/main/java/org/dsngroup/broke/broker/storage/InMemoryPool.java]: Yes. And ArrayList can be an efficient data structure for continuous data accessing.

File: broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessageParseHandler.java (hunk 1 above)
Comment (tz70s) [reply broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessageParseHandler.java]: oops, the **netty** message decoder/encoder mechanism.

File: broke-broker/src/main/java/org/dsngroup/broke/broker/storage/InMemoryPool.java (hunk 2 above)
Comment (tz70s) [reply broke-broker/src/main/java/org/dsngroup/broke/broker/storage/InMemoryPool.java]: I think we can discuss on this later b.c. the opinion of mine under.

**Comment by tz70s** [comment]
Totally GOOD try! I'll first merge this without modification is fine. We may jump to the next step for discussion and refactoring some parts.

**Comment by allenlee820202** [comment]
Thanks for the review!

**Pull request merged**